### PR TITLE
F/backend/phi

### PR DESCRIPTION
**Pull request by 0xNB**

### CODE REVIEW

File: src/main/java/edu/kit/compiler/codegen/BasicBlocks.java [hunk 1]
```diff
@@ -71,14 +72,14 @@ public int getLabel() {
          * Return the basic instructions of this block.
          */
         public List<Instruction> getInstructions() {
-            return Collections.unmodifiableList(instructions);
+            return instructions;
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/codegen/BasicBlocks.java:75]: There is a reason why this originally returned an unmodifiable list. You're not really supposed to change the state of this class. For example, I'm currently considering breaking `instructions` into two lists. You're change here makes that pretty much impossbile.

File: src/main/java/edu/kit/compiler/codegen/PhiRemover.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    public PhiRemover(BasicBlocks basicBlocks) {
+        this.basicBlocks = basicBlocks;
+    }
+
+    /**
+     * transforms this list of blocks into a new one without any phi instructions
+     *
+     * @return list of basic blocks without phis
+     */
+    public BasicBlocks removePhis() {
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/codegen/PhiRemover.java:35]: I would have expected your code to receive an instance of `BasicBlocks` and return some kind of `Map<Integer, List<Instruction>>` (which is what we need for register allocation).

File: src/main/java/edu/kit/compiler/codegen/PhiRemover.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                    PhiSourceNodeRegister targetRegister = new PhiSourceNodeRegister(phiInstruction.getTargetRegister(), nullNode);
+                    permutationSolver.addMapping(sourceRegister, targetRegister);
+                    int max = Math.max(source.getRegister(), phiInstruction.getTargetRegister());
+                    if (max > maxRegister) {
+                        maxRegister = max + 1;
+                    }
+                }
+            }
+            // solve permutations
+            List<PhiAssignment> permutations = permutationSolver.solve(new PhiSourceNodeRegister(maxRegister, nullNode));
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/codegen/PhiRemover.java:53]: You seem to be using `maxRegister` for spilling. However, as far as I can tell, there is no guarantee, that register isn't used elsewhere. I'm sure I can expose my register counter to you somehow, so you can properly allocate the temporary register.

**Review by g4rgoil — CHANGES_REQUESTED** [review]
Are currently dealing with the instruction generated by `ExitCondition` in some way?

**Comment by N-Maas** [comment]
Replaced by #103

**Pull request closed without merging**